storage/src/main/java/io/delta/storage/internal/ThreadUtils.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright (2021) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.delta.storage.internal;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.Callable;
+
+public final class ThreadUtils {
+
+    /**
+     * Based out of sparks ThreadUtils.runInNewThread
+     * Run a piece of code in a new thread and return the result.
+     * RuntimeException is thrown to avoid the calling interfaces
+     * from handling any Exceptions other than IOExceptions
+     */
+    public static <T> T runInNewThread(
+            String threadName,
+            boolean isDaemon,
+            Callable<T> body) throws IOException {
+        //Using a single element list to hold the exception and result,
+        //since T exception or T result cannot be used in static method
+        List<Exception> exceptionHolder = new ArrayList<>(1);
+        List<T> resultHolder = new ArrayList<>(1);
+        Thread thread = new Thread(threadName) {
+            @Override
+            public void run() {
+                try {
+                    resultHolder.add(body.call());
+                } catch (Exception ex) {
```

Comment:
We should catch `Throwable`. Otherwise, if an `Error` is thrown here, the caller thread will hang forever.